Tasks/Common/azure-blobstorage-artifactProvider/azureBlobStorageProvider.ts

```diff
@@ -0,0 +1,135 @@
+
+export class AzureBlobProvider implements models.IArtifactProvider {
+    constructor(storageAccount: string, container: string, accessKey: string, prefixFolderPath?: string, host?: string) {
+        this._storageAccount = storageAccount;
+        this._accessKey = accessKey;
+        this._container = container;
+        this._prefixFolderPath = prefixFolderPath;
+        this._blobSvc = azureStorage.createBlobService(this._storageAccount, this._accessKey, host);
+    }
+
+    public putArtifactItem(item: models.ArtifactItem, readStream: stream.Readable): Promise<models.ArtifactItem> {
+        return new Promise(async (resolve, reject) => {
+            var newArtifactItem: models.ArtifactItem = models.ArtifactItem.clone(item);
+            await this._ensureContainerExistence();
+
+            var self = this;
+            console.log(tl.loc("UploadingItem", item.path));
+            var blobPath = this._prefixFolderPath ? this._prefixFolderPath + "/" + item.path : item.path;
+
+            var writeStream = this._blobSvc.createWriteStreamToBlockBlob(this._container, blobPath, null, function (error, result, response) {
+                if (error) {
+                    console.log(tl.loc("FailedToUploadBlob", blobPath, error.message));
+                    reject(error);
+                } else {
+                    var blobUrl = self._blobSvc.getUrl(self._container, blobPath);
+                    console.log(tl.loc("CreatedBlobForItem", item.path, blobUrl));
+                    newArtifactItem.metadata["downloadUrl"] = blobUrl;
+                    resolve(newArtifactItem);
+                }
+            });
+
+            readStream.pipe(writeStream);
+            writeStream.on("error",
+                (error) => {
+                    console.log("ErrorInWriteStream", error);
+                    reject(error);
+                });
+            readStream.on("error",
+                (error) => {
+                    console.log(tl.loc("ErrorInReadStream", error));
+                    reject(error);
+                });
+        });
+    }
+
+    public getRootItems(): Promise<models.ArtifactItem[]> {
+        return new Promise(async (resolve, reject) => {
+            var items: models.ArtifactItem[] = await this._getItems(this._container)
+            resolve(items);
+        });
+    }
+
+    public getArtifactItems(artifactItem: models.ArtifactItem): Promise<models.ArtifactItem[]> {
+        return new Promise(async (resolve, reject) => {
+            var items: models.ArtifactItem[] = await this._getItems(this._container, artifactItem.path);
+            resolve(items);
+        });
+    }
+
+    public getArtifactItem(artifactItem: models.ArtifactItem): Promise<stream.Readable> {
+        return new Promise((resolve, reject) => {
+            var readStream: stream.Readable = this._blobSvc.createReadStream(this._container, artifactItem.path, null);
+            resolve(readStream);
```

Comment:
you should handle error.
Is the stream usable right away or do you need to process further in callback?

Reply:
further processing is not required in callback, infact if the promise is resolved inside callback then the readstream doesn't work properly.